### PR TITLE
CLI Command for "show-generation"

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch adds a new CLI command, _show-generation_, which displays a summary of the "next" generation consisting of applications with changes, units advanced to the generation, and configuration changes applied to the generation.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=generations`
- Bootstrap.
- `juju deploy redis -n 2`
- `juju add-generation`
- `juju advance-generation redis/0`
- `juju config redis databases=8`
- `juju show-generation`
- Output should match:
```yaml
next:
- application: redis
  units:
  - redis/0
  config:
    databases: 8
```

## Documentation changes

Yes, this CLI command should be documented with the generations feature generally.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Client-Side for "show-generation"](https://trello.com/c/tLaQdKfV/1118-client-side-for-show-generation)

**Comment by mitechie** [comment]
QA ok

File: api/modelgeneration/modelgeneration.go [hunk 1]
```diff
@@ -103,6 +104,37 @@ func (c *Client) HasNextGeneration(modelUUID string) (bool, error) {
...
+		return nil, errors.Trace(result.Error)
+	}
+	return generationInfoFromDTO(result.Applications), nil
+}
+
 func argForModel(modelUUID string) params.Entity {
 	return params.Entity{Tag: names.NewModelTag(modelUUID).String()}
 }
+
+func generationInfoFromDTO(apps []params.GenerationApplication) model.GenerationSummaries {
```
Comment (achilleasa) [inline api/modelgeneration/modelgeneration.go]: What does DTO stand for?

File: cmd/juju/model/showgeneration.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	showGenerationDoc     = `
+Summary information includes each application that has changes made under the
+generation, with any units that have been advanced, and any changed config
+values made to the generations.
+
+Examples:
+    juju add-generation
+
+See also:
+	add-generation
```
Comment (achilleasa) [inline cmd/juju/model/showgeneration.go]: Should this be left-aligned like the ones below it?

File: core/model/generation.go [hunk 3]
```diff
@@ -28,3 +28,24 @@ const generationKeySuffix = "#next"
...
+	// Config changes are the differing configuration values between this
+	// generation and the current.
+	// TODO (manadart 2018-02-22) This data-type will evolve as more aspects
+	// of the application are made generational.
+	ConfigChanges map[string]interface{} `yaml:"config"`
+}
+
+// GenerationSummaries is a type alias for a representation
+// of changes-by-generation.
+type GenerationSummaries = map[GenerationVersion][]GenerationApplication
```
Comment (achilleasa) [inline core/model/generation.go]: Is the `=` here valid Go or a typo?

File: api/modelgeneration/modelgeneration.go (hunk 1 above)
Comment (manadart) [reply api/modelgeneration/modelgeneration.go]: Data Transfer Object.

File: core/model/generation.go (hunk 3 above)
Comment (manadart) [reply core/model/generation.go]: That makes it a type alias as opposed to a new type.

File: cmd/juju/model/showgeneration.go (hunk 2 above)
Comment (manadart) [reply cmd/juju/model/showgeneration.go]: Yes; tabs vs spaces. Corrected.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**